source/extensions/common/wasm/wasm.cc

```diff
@@ -165,45 +160,87 @@ static void createWasmInternal(const VmConfig& vm_config, const PluginSharedPtr&
-    }
-    std::unique_ptr<Context> context(context_ptr);
-    auto configuration = vm_config.configuration();
-    auto vm_key = proxy_wasm::makeVmKey(vm_config.vm_id(), configuration, code);
-    cb(std::static_pointer_cast<WasmHandle>(proxy_wasm::createWasm(
-        vm_key, code, plugin,
-        [&vm_config, &configuration, &scope, &cluster_manager, &dispatcher,
-         &lifecycle_notifier](absl::string_view vm_key) {
-          auto wasm = std::make_shared<Wasm>(vm_config.runtime(), vm_config.vm_id(), configuration,
-                                             vm_key, scope, cluster_manager, dispatcher);
-          // NB: we need the shared_ptr to have been created for shared_from_this() to work.
-          wasm->initializeLifecycle(lifecycle_notifier);
-          return std::static_pointer_cast<WasmHandleBase>(std::make_shared<WasmHandle>(wasm));
-        },
-        vm_config.allow_precompiled(), std::move(context))));
-  };
+  auto complete_cb =
+      [cb, vm_config, plugin, scope, &cluster_manager, &dispatcher, &lifecycle_notifier,
+       root_context_for_testing_ptr = root_context_for_testing.release()](std::string code) {
+        auto root_context = std::unique_ptr<Context>(root_context_for_testing_ptr);
+        auto vm_key = proxy_wasm::makeVmKey(vm_config.vm_id(), vm_config.configuration(), code);
+        cb(std::static_pointer_cast<WasmHandle>(proxy_wasm::createWasm(
+            vm_key, code, plugin,
+            [&vm_config, &scope, &cluster_manager, &dispatcher,
+             &lifecycle_notifier](absl::string_view vm_key) {
+              auto wasm = std::make_shared<Wasm>(vm_config.runtime(), vm_config.vm_id(),
+                                                 vm_config.configuration(), vm_key, scope,
+                                                 cluster_manager, dispatcher);
+              // NB: we need the shared_ptr to have been created for shared_from_this() to work.
+              wasm->initializeLifecycle(lifecycle_notifier);
+              return std::static_pointer_cast<WasmHandleBase>(std::make_shared<WasmHandle>(wasm));
+            },
+            vm_config.allow_precompiled(), std::move(root_context))));
+      };
 
-  if (vm_config.code().has_remote()) {
+  if (fetch) {
+    // NB: if the (fetching) exception is thrown below, the remote_data provider will be deleted
+    // immediately rather than completing the async fetch, so allow for self-delete.
+    auto remote_data_provider_holder =
+        std::make_shared<std::unique_ptr<Config::DataSource::RemoteAsyncDataProvider>>();
+    auto fetch_callback = [vm_config, complete_cb, source, &dispatcher,
+                           remote_data_provider_holder](const std::string& code) {
+      {
+        std::lock_guard<std::mutex> guard(code_cache_mutex);
+        auto& e = (*code_cache)[vm_config.code().remote().sha256()];
+        e.in_progress = false;
+        e.code = code;
+      }
+      if (!fail_if_code_not_cached) {
```

Comment:
Is this so that we can still enter warming state?

Reply:
I have preserved the old behavior as an option.  It isn't clear to me how other people outside of istio might want to use this.  We can remove it later if there is no interest and if xDS doesn't provide a way to send back async status.